File: 36-valid-sudoku/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        // boxes
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                HashSet<Character> set = new HashSet<Character>();
                for (int k = 0; k < 3; k++) {
                    for (int l = 0; l < 3; l++) {
                        if (board[i * 3 + k][j * 3 + l] == '.')
                            continue;
                        if (set.contains(board[i * 3 + k][j * 3 + l]))
                            return false;
                        set.add(board[i * 3 + k][j * 3 + l]);
                    }
                }
            }
        }
        for (int i = 0; i < 9; i++) {
            HashSet<Character> set = new HashSet<Character>();
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.')
                    continue;
                if (set.contains(board[i][j]))
                    return false;
                set.add(board[i][j]);
            }
        }
        for (int i = 0; i < 9; i++) {
            HashSet<Character> set = new HashSet<Character>();
            for (int j = 0; j < 9; j++) {
                if (board[j][i] == '.')
                    continue;
                if (set.contains(board[j][i]))
                    return false;
                set.add(board[j][i]);
            }
        }
        return true;
    }
}

File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board):
        # Check 3x3 boxes
        for i in range(3):
            for j in range(3):
                seen = set()
                for k in range(3):
                    for l in range(3):
                        num = board[i * 3 + k][j * 3 + l]
                        if num == '.':
                            continue
                        if num in seen:
                            return False
                        seen.add(num)
        
        # Check rows
        for i in range(9):
            seen = set()
            for j in range(9):
                num = board[i][j]
                if num == '.':
                    continue
                if num in seen:
                    return False
                seen.add(num)
        
        # Check columns
        for i in range(9):
            seen = set()
            for j in range(9):
                num = board[j][i]
                if num == '.':
                    continue
                if num in seen:
                    return False
                seen.add(num)
        
        return True
